### PR TITLE
allow relabeling of alerts

### PR DESCRIPTION
**Pull request by brancz**
Regarding #1744

In case of dropping don't even enqueue them, that way dropped alerts don't interfere with the max batch size.

Tagging @brian-brazil as he originally created the issue.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -176,9 +176,10 @@ func (u URL) MarshalYAML() (interface{}, error) {
 
 // Config is the top-level configuration for Prometheus's config files.
 type Config struct {
-	GlobalConfig  GlobalConfig    `yaml:"global"`
-	RuleFiles     []string        `yaml:"rule_files,omitempty"`
-	ScrapeConfigs []*ScrapeConfig `yaml:"scrape_configs,omitempty"`
+	GlobalConfig        GlobalConfig     `yaml:"global"`
+	AlertRelabelConfigs []*RelabelConfig `yaml:"alert_relabel_configs,omitempty"`
```
Comment (brancz) [inline config/config.go]: open about discussions of naming and whether this belongs here

File: notifier/notifier.go [hunk 2]
```diff
@@ -231,6 +236,18 @@ func (n *Notifier) Send(alerts ...*model.Alert) {
 	n.setMore()
 }
 
+func (n *Notifier) relabelAlerts(alerts []*model.Alert) []*model.Alert {
+	var relabeledAlerts []*model.Alert
+	for _, alert := range alerts {
+		labels, _ := retrieval.Relabel(alert.Labels, n.opts.RelabelConfigs...)
```
Comment (fabxc) [inline notifier/notifier.go]: You wanted to remove the error from that function signature, right.
Let's move the relabel functionality into its own package while we are at it. `retrieval` has no logical relation with notifications.

File: notifier/notifier_test.go [hunk 3]
```diff
@@ -250,6 +260,13 @@ func TestHandlerFull(t *testing.T) {
 	h.Send(alerts[2*maxBatchSize : 3*maxBatchSize]...)
 	h.Send(alerts[3*maxBatchSize : 4*maxBatchSize]...)
 
+	// This alert should be dropped due to the configuration
+	h.Send(&model.Alert{
+		Labels: model.LabelSet{
+			"alertname": "drop",
+		},
+	})
+
```
Comment (fabxc) [inline notifier/notifier_test.go]: I'd vote for extracting this into its own test. Relabeling is merely an orthogonal pre-processing step, so this  mostly makes this test case harder to understand. And it's fairly complex to begin with.

For unit tests it's okay to have the actual code in mind. So how about a test function that applies one label modification and one dropping behavior and checks what's queued afterwards. All within the limit of a single batch.

We could rename this test to `TestHandlerQueueing` or similar.

File: notifier/notifier.go (hunk 2 above)
Comment (brancz) [inline notifier/notifier.go]: Yes you're right, I wanted to do that. Will move relabeling to it's own package, I think that makes sense as well.

File: notifier/notifier_test.go (hunk 3 above)
Comment (brancz) [inline notifier/notifier_test.go]: Sounds good to me

File: config/config.go (hunk 1 above)
Comment (fabxc) [inline config/config.go]: This is as we discussed.

But this is a good time to discuss opening a new top-level section called `alerts`. 
@brian-brazil, we need a section to handle service discovery for Alertmanagers. So I'd imagine we'd end up somewhere like:


```yaml
alerts:
  relabel_configs:
  - ...
  - ...
  alertmanagers:
    kubernetes_sd_config:
      ...
    relabel_configs:
    -  ...
```

I think we can save ourselves making lists of SDs in this case as well and should probably only allow using a single one, to be consistent.

(Why did we suffix everything with `_config` again in the past? Seems like bloat at best.)

Comment (brian-brazil) [inline config/config.go]: Yeah, a new section would be best. I'd go more for `alerting`, as users might think that `alerts` is where you define alerts.

I'd also call this `alert_relabel_configs` to reduce confusion.

File: notifier/notifier_test.go (hunk 3 above)
Comment (brancz) [inline notifier/notifier_test.go]: Done, can you check if it is as you imagined? @fabxc

File: config/config.go (hunk 1 above)
Comment (brancz) [inline config/config.go]: So I now changed it to have the top level `alerting` config node. That will make the config look like the following:

```
alerting:
  alert_relabel_configs:
  - action: ...
```

File: notifier/notifier_test.go [hunk 4]
```diff
@@ -185,6 +187,54 @@ func TestHandlerSendAll(t *testing.T) {
 	}
 }
 
+func TestHandlerQueuing(t *testing.T) {
+	dropRegex, _ := config.NewRegexp("drop")
+	renameRegex, _ := config.NewRegexp("rename")
```
Comment (fabxc) [inline notifier/notifier_test.go]: Never drop errors, in general but especially not in Go. Either check and `t.Fatalf` or use the `config.MustNewRegexp` constructur. The latter can directly be inlined in the options initialization below.

File: notifier/notifier_test.go [hunk 5]
```diff
@@ -185,6 +187,54 @@ func TestHandlerSendAll(t *testing.T) {
 	}
 }
 
+func TestHandlerQueuing(t *testing.T) {
```
Comment (fabxc) [inline notifier/notifier_test.go]: Ah, I was a bit unclear. I'd name this one `TestHandlerRelabel` and the existing on to this name.

File: relabel/relabel.go [hunk 6]
```diff
@@ -26,24 +26,20 @@ import (
 // Relabel returns a relabeled copy of the given label set. The relabel configurations
```
Comment (fabxc) [inline relabel/relabel.go]: Doc comment needs to be adjusted to new function name.

**Comment by fabxc** [comment]
That looks correct after comments.

File: notifier/notifier_test.go (hunk 5 above)
Comment (brancz) [inline notifier/notifier_test.go]: Ah ok wasn't exactly sure of it.

File: relabel/relabel.go (hunk 6 above)
Comment (brancz) [inline relabel/relabel.go]: will adjust

**Comment by brancz** [comment]
Alright I think I applied everything, one last rereview? :) @fabxc @brian-brazil

**Comment by fabxc** [comment]
We have a semantical quesiton open. We have instrumentation counting sent and dropped alerts. This is not adjusted for the ones dropped at relabeling.

I think we have to distinguish here. General `prometheus_notifications_dropped_total` is something I'd alert on at the moment, as it indicates queue saturation. Alerts dropped via relabeling are intentional though. So should we count them? And if yes, should they be their own metric?

**Comment by brancz** [comment]
Aha yes. I would say as it is typically an intentional behavior to drop alerts through relabeling a separate metric makes more sense. The `prometheus_notificaitons_dropped_total` is for notifications dropped due to technical reasons. I can imagine it can come in handy when users are unintentionally dropping alerts, if it's mixed with the count of notifications dropped due to queue size the reason becomes ambiguous.

**Comment by fabxc** [comment]
Agreed. I think for now it's fine then to leave it as is and add a new metric counting the alerts discarded based on relabeling.

👍

**Pull request merged**